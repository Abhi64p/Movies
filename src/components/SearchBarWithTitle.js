import React, {useEffect, useRef, useState} from 'react';
import {
  Image,
  StyleSheet,
  Text,
  TextInput,
  TouchableOpacity,
  View,
} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import Animated, {
  FadeInLeft,
  FadeInRight,
  FadeOutLeft,
  FadeOutRight,
} from 'react-native-reanimated';

import {Images} from '../assets/images';
import {Typography} from '../styles';

const {View: AView} = Animated;

const SearchBarWithTitle = ({title, onChangeText}) => {
  const textInputRef = useRef();
  const timeoutId = useRef();

  const [searchEnabled, setSearchEnabled] = useState(false);
  const [searchString, setSearchString] = useState('');

  useEffect(() => {
    if (searchEnabled) {
      textInputRef?.current?.focus?.();
    }
  }, [searchEnabled]);

  useEffect(() => {
    if (searchString === '' && !textInputRef?.current?.isFocused()) {
      setSearchEnabled(false);
    }

    /**
     * timeout is for handling debouncing
     */
    if (timeoutId.current) {
      clearTimeout(timeoutId.current);
    }
    timeoutId.current = setTimeout(() => onChangeText?.(searchString), 200);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchString]);

  const handleSearchStringChange = text => setSearchString(text);
  const handleClear = () => setSearchString('');
  const toggleSearch = () =>
    setSearchEnabled(!searchEnabled || searchString !== '');

  return (
    <>
      <View style={styles.container}>
        {searchEnabled ? (
          <AView
            key={`input${searchEnabled ? 'yes' : 'no'}`}
            style={styles.row0}
            entering={FadeInRight}
            exiting={FadeOutRight}>
            <TextInput
              ref={textInputRef}
              style={[styles.textInput, Typography.h3]}
              placeholder="Search movie title"
              placeholderTextColor="white"
              value={searchString}
              onChangeText={handleSearchStringChange}
              onBlur={toggleSearch}
            />
            <TouchableOpacity
              onPress={handleClear}
              disabled={searchString === ''}
              testID="searchStringClearButton">
              <Image
                source={Images.cancel}
                resizeMode="contain"
                style={styles.searchIcon}
              />
            </TouchableOpacity>
          </AView>
        ) : (
          <AView
            key={`title${searchEnabled ? 'yes' : 'no'}`}
            style={styles.flex1}
            entering={FadeInLeft}
            exiting={FadeOutLeft}>
            <Text style={Typography.h2}>{title}</Text>
          </AView>
        )}
        <TouchableOpacity
          style={styles.searchButton}
          onPress={toggleSearch}
          disabled={searchEnabled && searchString !== ''}
          testID="searchToggleButton">
          <Image
            source={Images.search}
            resizeMode="contain"
            style={styles.searchIcon}
          />
        </TouchableOpacity>
      </View>
      <LinearGradient
        colors={['rgba(32,33,36,1)', 'rgba(32,33,36,0)']}
        start={{x: 0, y: 0}}
        end={{x: 0, y: 1}}
        locations={[0, 1]}
        style={styles.gradient}
      />
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    marginHorizontal: 20,
    alignItems: 'center',
    marginTop: 10,
  },
  gradient: {height: 20, zIndex: 2},
  searchIcon: {height: 20, width: 20},
  searchButton: {marginLeft: 10, paddingVertical: 4},
  flex1: {flex: 1},
  textInput: {
    flex: 1,
    height: 28,
    padding: 0,
  },
  row0: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
    borderColor: 'white',
    borderWidth: 0.8,
    borderRadius: 3,
    paddingLeft: 8,
    paddingRight: 5,
  },
});

export default SearchBarWithTitle;
