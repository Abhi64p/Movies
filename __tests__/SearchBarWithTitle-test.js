import React from 'react';
import {render, fireEvent, waitFor} from '@testing-library/react-native';
jest.mock('react-native-linear-gradient', () => 'LinearGradient');
jest.mock('react-native-reanimated', () => ({View: 'AView'}));

import SearchBarWithTitle from '../src/components/SearchBarWithTitle';

test('Test to see if title is rendering', () => {
  const TITLE = 'title for testing';
  const {getByText, toJSON} = render(<SearchBarWithTitle title={TITLE} />);
  const titleElement = getByText(TITLE);
  expect(titleElement).toBeDefined();
  expect(toJSON()).toMatchSnapshot();
});

test('Test to see if Title <-> SearchBar toggle is working', () => {
  const TITLE = 'title for testing';
  const {getByText, getByPlaceholderText, getByTestId} = render(
    <SearchBarWithTitle title={TITLE} />,
  );
  const searchToggleButton = getByTestId('searchToggleButton');
  fireEvent(searchToggleButton, 'onPress'); // Switching to Search bar from Title
  const textInput = getByPlaceholderText('Search movie title');
  expect(textInput).toBeDefined();
  fireEvent(searchToggleButton, 'onPress'); // Switching back to Title
  const titleElement = getByText(TITLE);
  expect(titleElement).toBeDefined();
});

test('Test to see if searching works', async () => {
  const STRING = 'random string';
  const onChangeText = jest.fn();
  const {getByPlaceholderText, getByTestId} = render(
    <SearchBarWithTitle onChangeText={onChangeText} />,
  );
  const searchToggleButton = getByTestId('searchToggleButton');
  fireEvent(searchToggleButton, 'onPress');
  const textInput = getByPlaceholderText('Search movie title');
  fireEvent.changeText(textInput, STRING);
  await waitFor(() => expect(onChangeText).toBeCalledWith(STRING));
});
