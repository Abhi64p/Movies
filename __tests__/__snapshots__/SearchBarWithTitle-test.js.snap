// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to see if title is rendering 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginHorizontal": 20,
        "marginTop": 10,
      }
    }
  >
    <AView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontFamily": "TitilliumWeb-Light",
            "fontSize": 20,
          }
        }
      >
        title for testing
      </Text>
    </AView>
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": 10,
          "opacity": 1,
          "paddingVertical": 4,
        }
      }
      testID="searchToggleButton"
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../src/assets/images/search.png",
          }
        }
        style={
          Object {
            "height": 20,
            "width": 20,
          }
        }
      />
    </View>
  </View>,
  <LinearGradient
    colors={
      Array [
        "rgba(32,33,36,1)",
        "rgba(32,33,36,0)",
      ]
    }
    end={
      Object {
        "x": 0,
        "y": 1,
      }
    }
    locations={
      Array [
        0,
        1,
      ]
    }
    start={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    style={
      Object {
        "height": 20,
        "zIndex": 2,
      }
    }
  />,
]
`;
